docker.Makefile

```diff
@@ -10,7 +10,11 @@ LINTER_IMAGE_NAME = docker-cli-lint$(IMAGE_TAG)
 CROSS_IMAGE_NAME = docker-cli-cross$(IMAGE_TAG)
 VALIDATE_IMAGE_NAME = docker-cli-shell-validate$(IMAGE_TAG)
 E2E_IMAGE_NAME = docker-cli-e2e$(IMAGE_TAG)
+GO_BUILD_CACHE ?= y
 MOUNTS = -v "$(CURDIR)":/go/src/github.com/docker/cli
+ifeq ($(GO_BUILD_CACHE),y)
+MOUNTS += -v "$(CURDIR)/.go-build-cache:/root/.cache/go-build"
```

Comment:
Why not use an (anonymous) volume instead of a bind-mount?

Also wondering if we should start considering _not_ bind-mounting the source, but rebuild the image instead, with BuildKit being optimized for rebuilding. In that case, we can use the `RUN --mount` option for the cache;

```Dockerfile
# syntax=docker/dockerfile:1.0-experimental
RUN --mount=target=. --mount=type=cache,target=/go/src/github.com/docker/cli/.cache/go-build ....
```

Reply:
> Why not use an (anonymous) volume instead of a bind-mount?

Won't an anonymous volume be freshly created for each build, rendering the cache useless? A named volume would work but do we want to be creating those here? A local directory seems simpler and easier to manage (e.g. the `clean` target can easily deal with it without the potential for getting confused by a users other volumes etc)

re buildkit, we should probably consider it but it would be a more major revamping of the build so should be a separate PR IMHO.

Reply:
> Won't an anonymous volume be freshly created for each build

Yes, it would; otoh (I think) the usecase here is to rebuild while inside the container, so the penalty for a "fresh" cache would only be once.

> A local directory seems simpler and easier to manage

Yes, and no; it would pollute my local machine

> without the potential for getting confused by a users other volumes etc

I guess if we carefully pick a name for the volume, that shouldn't be a big issue; here's what we do in moby/moby https://github.com/moby/moby/blob/master/Makefile#L75

```Makefile
DOCKER_MOUNT_CACHE := -v docker-dev-cache:/root/.cache
```

Reply:
> Yes, it would; otoh (I think) the usecase here is to rebuild while inside the container, so the penalty for a "fresh" cache would only be once.

As I said in the intro the use case is running `make build -f docker.Makefile` repeatedly. It's good that it also helps with `make shell -f docker.Makefile` though.

I'll switch to use a named `docker-cli-dev-cache` volume.